File: 11_DataStruct_and_Algorithm/DataStruct/Practice/binarySearch.cpp
template <typename T>
int binarySearch(T a[], int n, const T& x) {
    // Search x in a[0:n-1].
    int left = 0;
    int right = n - 1;
    while (left <= right) {
        int middle = (left + right) / 2;
        if (x == a[middle]) return middle;
        if (x > a[middle]) left = middle + 1;
        else right = middle - 1;
    }
    return -1;
}

File: 2_Computer_Science/Artificial_Intelligence/experiment/EXP1/BinarySearch.py
def BinarySearch(nums, target):
    """
    :param nums: list[int]
    :param targe: int
    :return int
    """

    left, right = 0, len(nums) - 1

    while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1


if __name__ == "__main__":
    nums = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    target = 5
    print(BinarySearch(nums, target))  # 4
    target = 10
    print(BinarySearch(nums, target))  # -1